### PR TITLE
correctly show license modified on new works

### LINKED ISSUE #2686: When new item is saved without any changes, history entry should not indicate any changes
If I create a new deposit and save it as a draft without making any changes, the history says "license modified" when it should not say anything.
Screen Shot 2022-09-14 at 11 10 22 AM

If there is a new item in a collection that lets a depositor choose the license and they choose a license other than the default, show the "license modified", if they use the default license then show no message.

### PR DESCRIPTION
**Pull request by peetucket**
## Why was this change made? 🤔

Fixes #2686 - indicate if a license was changed from the default for a brand new work; if not, keep the current message ("Created")

Todo: 
- [x] add a test
- [x] update functionality for how events work (see ticket comments)

## How was this change tested? 🤨

Localhost and added a request test

### CODE REVIEW

File: spec/requests/create_work_spec.rb [hunk 1]
```diff
@@ -367,20 +368,52 @@
...
+            expect(work_version.contact_emails.size).to eq 1
+            expect(work_version.published_edtf).to be_nil
+            expect(work_version.created_edtf).to be_nil
+            expect(work_version.embargo_date).to be_nil
+            expect(work_version.subtype).to be_empty
+            expect(work_version.state).to eq 'depositing'
+            expect(DepositJob).to have_received(:perform_later)
+            expect(last_event.event_type).to eq 'update_metadata'
+            expect(last_event.description).to eq 'Created'
+          end
```
Comment (peetucket) [inline spec/requests/create_work_spec.rb:392]: the new expectation for this existing test

File: spec/requests/create_work_spec.rb [hunk 2]
```diff
@@ -367,20 +368,40 @@
...
+          let(:license) { 'Apache-2.0' }
+
+          it 'indicates correct event type as license modified' do
+            post "/collections/#{collection.id}/works", params: { work: work_params, commit: 'Deposit' }
+            expect(response).to have_http_status(:found)
+            last_event = Work.last.events.last
+            expect(last_event.event_type).to eq 'update_metadata'
+            expect(last_event.description).to eq 'license modified'
+          end
         end
```
Comment (peetucket) [inline spec/requests/create_work_spec.rb:405]: a new test

**Comment by amyehodge** [comment]
Here's my thinking on what should be happening:

IF an item is created and saved without anything being done in the form: “Created”
IF an item is created and the license is changed (and nothing else): “License updated”
IF an item is created and the license is changed and other things are changed: “License changed, title modified, keywords modified, etc.”

**Comment by peetucket** [comment]
> Here's my thinking on what should be happening:
> 
> IF an item is created and saved without anything being done in the form: “Created” IF an item is created and the license is changed (and nothing else): “License updated” IF an item is created and the license is changed and other things are changed: “License changed, title modified, keywords modified, etc.”

Ok, so this will require some more rethinking and work then.

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -26,7 +26,7 @@ def create
 
     @form = work_form(work_version)
     if @form.validate(work_params) && @form.save
-      after_save(form: @form, event_context: { user: current_user, description: 'Created' })
+      after_save(form: @form, event_context: build_event_context(form: @form, event_type: :create))
     else
```
Comment (peetucket) [inline app/controllers/works_controller.rb:30]: We want to record changes for a new work, but the logic is bit different.  That logic will be handled in the class, but we need to tell it this is a new work.

File: app/services/work_version_event_description_builder.rb [hunk 4]
```diff
@@ -1,24 +1,32 @@
 # frozen_string_literal: true
 
-# This creates an appropriate description for an update event on a WorkVersion
+# This creates an appropriate description for a create/update event on a WorkVersion
 class WorkVersionEventDescriptionBuilder
```
Comment (peetucket) [inline app/services/work_version_event_description_builder.rb:4]: this class will change to handle the new differing logic for new vs updated works

File: spec/requests/create_work_spec.rb [hunk 5]
```diff
@@ -418,6 +424,85 @@
...
+            expect(response).to have_http_status(:found)
+            work_version = Work.last.head
+            expect(work_version.license).to eq license
+            last_event = Work.last.events.last
+            expect(last_event.event_type).to eq 'update_metadata'
+            expect(last_event.description).to eq 'license modified'
+          end
+        end
+      end
+
```
Comment (peetucket) [inline spec/requests/create_work_spec.rb:505]: new tests that verify that a created work gets the correct event description for a few scenarios (no changes, default license changed, title changed)

File: spec/services/work_version_event_description_builder_spec.rb [hunk 6]
```diff
@@ -3,422 +3,468 @@
 require 'rails_helper'
 
 RSpec.describe WorkVersionEventDescriptionBuilder do
-  subject(:result) { described_class.build(form) }
+  subject(:result) { described_class.build(form:, event_type:) }
```
Comment (peetucket) [inline spec/services/work_version_event_description_builder_spec.rb:7]: This file didn't change as much as the diffs suggest.  Basically I put in two context blocks, one for all of the existing tests (representing an update to an existing work) that made all of the existing tests become indented, and a new context block for new tests related to the creation of a new work.

File: app/controllers/works_controller.rb [hunk 7]
```diff
@@ -26,7 +26,7 @@ def create
 
     @form = work_form(work_version)
     if @form.validate(work_params) && @form.save
-      after_save(form: @form, event_context: { user: current_user, description: 'Created' })
+      after_save(form: @form, event_context: build_event_context(@form))
     else
```
Comment (peetucket) [inline app/controllers/works_controller.rb:30]: it turns out we *do* want to show changes for a new work after all, but with some slightly different logic (which will be handled in the event builder class)

File: spec/features/create_new_work_spec.rb [hunk 8]
```diff
@@ -162,8 +163,11 @@
...
 
           within '#events' do
-            # New work gets a description of "Created"
-            expect(page).to have_content 'Created', count: 1
+            # New work gets a description of what's changed (except subtypes)
+            changed = ['title of deposit modified', 'abstract modified', 'contact email modified', 'authors modified',
+                       'publication date modified', 'creation date modified', 'keywords modified', 'citation modified',
+                       'files added/removed']
+            expect(page).to have_content changed.join(', '), count: 1
           end
```
Comment (peetucket) [inline spec/features/create_new_work_spec.rb:171]: new works once again get the changes indicated

File: spec/features/create_new_work_spec.rb [hunk 9]
```diff
@@ -382,8 +386,11 @@
...
```
Comment (peetucket) [inline spec/features/create_new_work_spec.rb:394]: new works once again get the changes indicated

File: spec/requests/create_work_spec.rb [hunk 10]
```diff
@@ -381,6 +381,12 @@
...
           expect(work_version.state).to eq 'depositing'
           expect(DepositJob).to have_received(:perform_later)
+          last_event = Work.last.events.last
+          expect(last_event.event_type).to eq 'update_metadata'
+          changes = ['title of deposit modified', 'abstract modified', 'contact email modified', 'authors modified',
+                     'keywords modified', 'visibility modified', 'license modified', 'files added/removed',
+                     'file description changed']
+          expect(last_event.description).to eq changes.join(', ')
         end
       end
```
Comment (peetucket) [inline spec/requests/create_work_spec.rb:391]: some additional expectations for changes on new works

**Review by jcoyne — APPROVED** [review]

**Pull request merged**